agent/agent_endpoint.go

```diff
@@ -1487,3 +1490,90 @@ func (s *HTTPServer) AgentHost(resp http.ResponseWriter, req *http.Request) (int
 
 	return debug.CollectHostInfo(), nil
 }
+
+// AgentBootstrapGossipKey
+//
+// PUT /v1/agent/bootstrap-gossip-key
+//
+// Allows delivering the gossip encryption key via API to avoid it being present
+// on disk if -encrypt-key-from-api and -disable-keyring-file are used.
+func (s *HTTPServer) AgentBootstrapGossipKey(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	// Fetch the ACL token, if any, and enforce agent policy. We require
+	// AgentWrite since we are effectively allowing the agent to join the cluster
+	// by delivering it's key.
+	//
+	// TODO(banks): is this necessary? Seems safer to be more restrictive rather
+	// than less for now. It seems the worst an attacker could do if unprotected
+	// would be to force the agent to join some other gossip pool they have setup
+	// but that could be a bad thing if there are multiple Consul clusters in the
```

Comment:
If you could interact with the local agent in such a way as to cause an OOM/crash/restart, then you could use this api to turn off gossip completely for the node (with your own random key) until an operator restarts and reinjects the gossip key.

Reply:
How would you fix the random key it uses? We read from crypto.rand on every process start for the temporary key?

Reply:
I meant you can neuter the agent by:
1. force OOM+restart by $magic
2. detect that consul has restarted and invoke the new API yourself with a key the attacker makes up (but before the local orchestrator has a chance to inject the correct one)
3. Now the orchestrator cannot use the new api, as a key has already been set.

This is mostly a DOS vector as described above.